libs/filer/files_client.go

```diff
@@ -165,40 +197,91 @@ func (w *FilesClient) Delete(ctx context.Context, name string, mode ...DeleteMod
 		return CannotDeleteRootError{}
 	}
 
-	err = w.apiClient.Do(ctx, http.MethodDelete, urlPath, nil, nil, nil)
-
-	// Return early on success.
-	if err == nil {
-		return nil
+	info, err := w.Stat(ctx, name)
+	if err != nil {
+		return err
 	}
 
-	// Special handling of this error only if it is an API error.
-	var aerr *apierr.APIError
-	if !errors.As(err, &aerr) {
+	if info.IsDir() {
+		err = w.workspaceClient.Files.DeleteDirectoryByDirectoryPath(ctx, absPath)
+
+		// This API returns a 400 if the directory is not empty
+		var aerr *apierr.APIError
+		if errors.As(err, &aerr) && aerr.StatusCode == http.StatusBadRequest {
+			return DirectoryNotEmptyError{absPath}
```

Comment:
I'm not sure you can assume that if the API return 400 it means that the directory is not empty.

For example, if the user input is bad, it could also return a 400. Maybe you could rely on the error reason set in the payload?

Reply:
This is documented in the API docs: https://openapi.dev.databricks.com/api/azure/workspace/files/deletedirectory

I agree this is not great though. Maybe `409` is a better error code to send.